src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package remote
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/generated/proto/rpc"
+
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3db/src/dbnode/ts"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+	"github.com/m3db/m3db/src/dbnode/x/xpool"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+	xtime "github.com/m3db/m3x/time"
+)
+
+// RPCFromSeriesIterator converts a seriesIterator to a grpc series
+func RPCFromSeriesIterator(it encoding.SeriesIterator) (*rpc.Series, error) {
+	defer it.Close()
+
+	replicas := it.Replicas()
+	compressedReplicas := make([]*rpc.CompressedValuesReplica, 0, len(replicas))
+	for _, replica := range replicas {
+		replicaSegments := make([]*rpc.Segments, 0, len(replicas))
+		readers := replica.Readers()
+		next := true
+		for next {
+			segments := &rpc.Segments{}
+			l, _, _ := readers.CurrentReaders()
+			if l == 1 {
+				br := readers.CurrentReaderAt(0)
+				segment, err := br.Segment()
+				if err != nil {
+					return nil, err
+				}
+				segments.Merged = &rpc.Segment{
+					Head: segment.Head.Bytes(),
+					Tail: segment.Tail.Bytes(),
+				}
+			} else {
+				unmerged := make([]*rpc.Segment, 0, l)
+				for i := 0; i < l; i++ {
+					br := readers.CurrentReaderAt(i)
```

Comment:
This section seems similar to the one above. Does it make sense to extract this logic out into a function?